### ISSUE TITLE
Is monetdbe threadsafe?

### ISSUE THREAD

**Issue by krimp** [MonetDBSolutions/monetdbe-examples#11]
I am testing MonetDBe in a multi threaded application (C++/Linux), where some threads insert data into the database and other threads read.

In your blog about MonetDBe you claim that MonetDBe is thread safe. However, for the time being it seems to me that I have to protect the database access by using a mutex to avoid segfault (testing with in-memory database, i.e no persistence).

Do you have any thoughts about that? Are there any implementation guidelines to follow for using MonetDBe in a threaded application where different threads access the in-memory database? I am using a global variabel monetdbe_database _db referencing the opened in-memory database (and hence accessible for all threads). 

A sort of pseudo code (sorry about the messy formatting):

`	
	monetdbe_database _dbe = NULL;
        mutex monetdbeMutex;
	bool _sysRunning = false;
	
	int monetDBeCreateDB(){
		monetdbe_options dbOptions;
		dbOptions.memorylimit    = 1000;
		dbOptions.querytimeout   = 60;
		dbOptions.sessiontimeout = 60; 
		dbOptions.nr_threads     = 4;

		char *msg= NULL;

		int statusDB = monetdbe_open(&_dbe, NULL /*pathDB*/, &dbOptions);
		if (statusDB < 0){
			return statusDB;
		} 

		msg = monetdbe_query(_dbe, "CREATE TABLE table0 (t bigint, x0 double, x1 double, x2 double, x3 double, x4 double, x5 double, x6 double);", NULL, NULL);
		if (msg != NULL) printf("%s\n", msg);

		msg = monetdbe_query(_dbe, "CREATE TABLE table1 (t bigint, y0 double, y1 double, y2 double, y3 double);", NULL, NULL);
		if (msg != NULL) printf("%s\n", msg);

		msg = monetdbe_query(_dbe, "CREATE TABLE table2   (t bigint, y0 double, y1 double, y2 double, y3 double);", NULL, NULL);
		if (msg != NULL) printf("%s\n", msg);

		msg = monetdbe_query(_dbe, "CREATE TABLE table3    (t bigint, u0 double, u1 double, u2 double, u3 double, u4 double);", NULL, NULL);
		if (msg != NULL) printf("%s\n", msg);

		return statusDB;
	}

	void thread0(){
	  
	   char* msg = NULL;
	   monetdbe_result *result;

	   while (_sysRunning){
			this_thread::sleep_for(chrono::milliseconds(10000));
	    	if (_dbe != NULL){
			    monetdbeMutex.lock();
			    msg = monetdbe_query(_dbe, "SELECT sys.epoch(t) as t_utc,* FROM table0 order by t desc limit 10;", &result, NULL);
				monetdbeMutex.unlock();

				if ( msg == NULL) {
					//do something with the data

				}

			    monetdbeMutex.lock();
				monetdbe_cleanup_result(_dbe, result);
				monetdbeMutex.unlock();
			}
		}
	}

	void thread1(){
	  
	   char* msg = NULL;
	   string dbQueryStr;
	   int64_t rows_affected;  

	   while (_sysRunning){
	        this_thread::sleep_for(chrono::milliseconds(1000));
	    	if (_dbe != NULL){
				dbQueryStr  = "INSERT INTO table0 (t,x0,x1,x2,x3,x4,x5,x6) VALUES (" ... some values ... ");"; 	
				dbQueryStr += "INSERT INTO table1 (t,y0,y1,y2,y3) VALUES ("        ... som values ... ");"; 	            	
				dbQueryStr += "INSERT INTO table2 (t,u0,u1,u2,u3,u4) VALUES ("      ... some values ... ");"; 	            	
				dbQueryStr += "INSERT INTO table3 (t,y0,y1,y2,y3) VALUES (" ... some values ... ");"; 	   

				monetdbeMutex.lock();
				msg = monetdbe_query(_dbe, (char *)dbQueryStr.c_str(), NULL, &rows_affected); 
				monetdbeMutex.unlock();

				if (msg != NULL) {
		                        printf("%s\n",dbQueryStr.c_str());
					printf("%s\n", msg);
				}
	    	}
	    } //while
	}

	int main(int argc, char **ppArgv){

		if (monetDBeCreateDB() >= 0){
			_sysRunning = true;     //must be set false by e.g. a sighandler to stop the application
			threads.push_back(std::thread(thread0)); 
			threads.push_back(std::thread(thread1)); 
		}

	 	while (_sysRunning){
			this_thread::sleep_for(chrono::milliseconds(60000));
	 	}
	        if (_dbe != NULL){
 	             monetdbe_close(_dbe);
   	            _dbe = NULL;
              }
	}
`

In the code like above, if the mutex protection are removed, the application segfaults after a while.

Tested on RPi4B, g++.

**Comment by MitchellWeg** [comment]
I've converted your example into C and can confirm it segfauts.
We're currently working on it.

**Comment by aris-koning** [comment]
Hi,

Yes, monetdbe is thread-safe if concurrent access is correctly implemented. Each concurrent client-side thread should operate on its own non-shared `monetdbe_database` handle. Admittedly we were missing an example demonstrating concurrent use. So we have just added [this example](https://github.com/MonetDBSolutions/monetdbe-examples/blob/master/C/concurrent.c) to show how this can be done using pthreads.

In your code (without the mutex) the different threads that are calling the monetdbe API are _concurrently_ sharing the `monetdbe_database` handle. We don't support that use.

We have also extended the documentation a bit here under [porting C programs](https://github.com/MonetDBSolutions/MonetDBe-Python/blob/master/doc/migrations.rst). It will be visible in [the monetdbe documentation portal](https://monetdbe.readthedocs.io) soon.
